spring-cloud-sleuth-samples/spring-cloud-sleuth-sample-messaging/src/test/java/integration/MessagingApplicationTests.java

```diff
@@ -99,14 +103,56 @@ private void thenAllSpansHaveTraceIdEqualTo(long traceId) {
 		then(this.integrationTestSpanCollector.hashedSpans.stream().allMatch(span -> span.traceId == traceId)).isTrue();
 	}
 
-	private void thenTheLastSpansParentHasIdEqualToFirstSpansId() {
-		Optional<Span> firstSpan = this.integrationTestSpanCollector.hashedSpans.stream()
-				.filter(span -> "http/".equals(span.name) && span.parentId != null).findFirst();
-		Optional<Span> lastSpan = this.integrationTestSpanCollector.hashedSpans.stream()
+	private void thenTheSpansHaveProperParentStructure() {
+		Optional<Span> firstHttpSpan = findFirstHttpRequestSpan();
+		List<Span> eventSpans = findAllEventRelatedSpans();
+		Optional<Span> eventSentSpan = findSpanWithServerSentAnnotation(eventSpans);
+		Optional<Span> eventReceivedSpan = findSpanWithClientReceivedAnnotation(eventSpans);
+		Optional<Span> lastHttpSpan = findLastHttpSpan();
+		thenAllSpansArePresent(firstHttpSpan, eventSpans, lastHttpSpan, eventSentSpan, eventReceivedSpan);
+		then(lastHttpSpan.get().parentId).isEqualTo(eventSentSpan.get().id);
+		then(eventSentSpan.get().parentId).isEqualTo(firstHttpSpan.get().id);
+		then(eventSentSpan.get()).isNotEqualTo(eventReceivedSpan.get());
+	}
+
+	private Optional<Span> findLastHttpSpan() {
+		return this.integrationTestSpanCollector.hashedSpans.stream()
 				.filter(span -> "http/foo".equals(span.name)).findFirst();
-		then(firstSpan.isPresent()).isTrue();
-		then(lastSpan.isPresent()).isTrue();
-		then(lastSpan.get().parentId).isEqualTo(firstSpan.get().id);
+	}
+
+	private Optional<Span> findSpanWithClientReceivedAnnotation(List<Span> eventSpans) {
```

Comment:
this logic could be deduped by adding a param for the annotation value. I think it would read almost as well.

```
findSpanWithServerSentAnnotation(eventSpans)
findSpanWithAnnotation(eventSpans, SERVER_SENT)
```